Fix #101: Remove SimpleType

ttreyer opened the pull request:
`SimpleType` has been removed from sbt and a test has been added to show that projection on structure are correctly invalidated by sbt now.

jvican committed:
Replace shapeless by scalac

ttreyer committed:
Add test for Projection on Structure type

ttreyer committed:
Fix #101: Remove SimpleType

lightbend-cla-validator commented:
Hi @zehkae,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

jvican commented on internal/zinc-apiinfo/src/main/scala/xsbt/api/SameAPI.scala:
```diff
@@ -259,13 +259,17 @@ class SameAPI(includePrivate: Boolean, includeParamNames: Boolean) {
     samePending(a, b)(sameTypeDirect)
   def sameTypeDirect(a: Type, b: Type): Boolean =
```
Can we enclose this complex expression with `{` and `}`?

jvican commented on zinc/src/sbt-test/source-dependencies/struct-projection/test:
```diff
@@ -0,0 +1,4 @@
+> compile
+$ copy-file changes/A.scala A.scala
```
Great test, I like it :wink:.

jvican reviewed (approved):
Thanks for the contribution @zehkae!
  
Good job, I like this simplification. It's something that had to be done sooner or later.
  
I've just asked for a minor change, otherwise it looks good to me! Can you perhaps execute `runBenchmarks` on a controlled environment? I'd like to see the results. For this, make sure the cpu frequency is fixed and the governor is `performance`. Have a look at the [CONTRIBUTING  guide](https://github.com/sbt/zinc/blob/1.0/CONTRIBUTING.md).

dwijnand reviewed (approved):
Thanks, @zehkae. Looks good.

@jvican I'm not sure what process you want to use, but please note that this is a breaking change, for the migration notes.

jvican commented:
@dwijnand This is not a breaking change because this API never got exposed to the public that I know of, it's internal to the compiler bridge.

ttreyer commented:
I made the change you asked for by amending the last commit.
I'm not yet used to GitHub's Pull Request, so please tell me if it was incorrect to use `--amend`.

I ran the benchmark on my laptop as follows:
```sh
systemctl stop ondemand.service
cpupower frequency-set --min 1.5GHz
cpupower frequency-set --max 1.5GHz
cpupower frequency-set --governor performance
sbt runBenchmarks
```
The results are [here](https://github.com/sbt/zinc/files/829242/benchmark.txt).

Also, since your bot asked me to tell: I signed the CLA :smiley:

jvican commented:
@zehkae Thanks! Can you run it before and after the benchmark? We need to compare if something has changed and I cannot do it because the environment won't be the same :wink:.

eed3si9n commented:
I agree that it's not a breaking change to the public Zinc API, but note that the API info does get persisted into gzip file, and is extracted by sbt to do things like test detection.

dwijnand commented:
@jvican you're right. I've become so hyper-sensitive to breaking changes that I hadn't realised this wasn't public API.

ttreyer commented:
@jvican [Here](https://github.com/sbt/zinc/files/830614/benchmark-before.txt) are the benchmarks before the patch, and [here](https://github.com/sbt/zinc/files/830618/benchmark-after.txt) are the results after.

jvican commented:
@zehkae This change has an unexpected hit in performance. You've run the benchmarks at different times (since you've reused the benchmark of after from yesterday) and that may be the culprit of that hit. I'll run this on my computer and see how it affects it.

jvican commented on internal/zinc-apiinfo/src/main/scala/xsbt/api/SameAPI.scala:
```diff
@@ -258,14 +258,20 @@ class SameAPI(includePrivate: Boolean, includeParamNames: Boolean) {
+        case (pa: Projection, pb: Projection)       => debug(sameProjection(pa, pb), "Different projection")
+        case (pa: ParameterRef, pb: ParameterRef)   => debug(sameParameterRef(pa, pb), "Different parameter ref")
+        case (sa: Singleton, sb: Singleton)         => debug(sameSingleton(sa, sb), "Different singleton")
+        case (_: EmptyType, _: EmptyType)           => true
```
This should be the last item in the match since it's not likely to happen. C2/C1 may reorder it based on hit frequency, but cold compilation will probably not see this optimization, so it's better to do it by hand.

jvican commented on internal/zinc-apiinfo/src/main/scala/xsbt/api/SameAPI.scala:
```diff
@@ -258,14 +258,20 @@ class SameAPI(includePrivate: Boolean, includeParamNames: Boolean) {
+        case (aa: Annotated, ab: Annotated)         => debug(sameAnnotatedType(aa, ab), "Different annotated types")
+        case (sa: Structure, sb: Structure)         => debug(sameStructureDirect(sa, sb), "Different structure type")
+        case (ea: Existential, eb: Existential)     => debug(sameExistentialType(ea, eb), "Different existential type")
+        case (pa: Polymorphic, pb: Polymorphic)     => debug(samePolymorphicType(pa, pb), "Different polymorphic type")
```
Can we move polymorphic higher in the hierarchy? Third position would be good.

dwijnand replied:
oh the opportunities to backdoor-condition what style of Scala people write, by micro-optimising these parts of the incremental compiler. I'm sure the same opportunities arise in the compiler itself actually.

smarter replied:
@jvican did your benchmarking show that we spend a significant amount of time in this method? Otherwise there is no point in optimizing it at all.

jvican replied:
It's not even about optimizing, it's about logical clarity. Having polymorphic be so at the end of the logic line is not good, especifically because how important and frequent it is. Leaving aside the microoptimization issues, it makes no sense to have polymorphic after annotated, structure and existential types IMO.

jvican commented:
Hey @zehkae, I'm benchmarking this PR today, will report you the results soon :wink:.

eed3si9n commented:
What's the status on this PR? Are we still waiting on some improvements?

jvican commented:
I was doing benchmarks but this slipped through the cracks. I'll post results today.

jvican commented:
Sorry for the slowliness in this PR @zehkae, but there's something that is not right. The measurements in my machine after this patch have too much fluctuation and show an increase in running time similar to the one you reported. I'll try to pin threads to get more predictable results and identify whether this decrease is real and why happens.
  
In the meanwhile, can you run benchmarks again? Use the Scala standard library and repeat the steps to set cpu frequency and governor. Make sure that you are not running any other processes while you benchmark and that they run exactly one after the other.

jvican committed:
Add Java Thread affinity for benchmarks

To avoid excessive fluctuation in benchmarks, we use Java Thread
affinity that help us pin the benchmarks to concrete cores/cpus.

jvican committed:
Reorder cases in type patmat

Reorder the cases not because it can cause a runtime overhead (which it
can) but because of mental clarity. Most frequent types should always be
at the top so that the reader is not distracted with the uninteresting
cases.

jvican commented:
I have pushed two commits: one that reorders the statements (so that you don't have to do it) and the other one that adds Java Thread Affinity for the benchmarks. Please, update your master and try running the benchmarks again with JTA enabled. Make sure that JNA (Java Native Access) is installed in your computer first.
  
In the meanwhile, I'll run them, but please run them again. I'm curious to see the results.

jvican commented:
Oh, and by the way, don't forget to cherry-pick the JTA commit whenever you want to test the results *before* this patch.

smarter commented:
@jvican I think you're too stringent. You shouldn't require external contributors to run benchmarks for hours on their computer without being able to do anything else. Furthermore, the benchmark results posted so far show that the difference is within the margin of error so I don't see what you hope to learn here. There are way too many factors at play here to be able to understand microscopic difference. If you want good benchmarking results, I suggest that you setup a CI that reports results for each merged PR and shows a nice graph like http://arewefastyet.com/, this will provide much better information than benchmarks run on contributors' machines.

jvican commented:
This is a pretty intrusive change in Zinc. I think I'm not being too stringent. This software is going to be used in production real soon, so I don't want any performance regression to happen, and even less in a PR that only addresses *an issue of correctness*, it does not add anything to the current status of Zinc. I'm not happy to sacrifice 700 ms with a duplicated margin error.
  
Benchmarks in CI are even less reliable than running them in a computer, especifically if you have no control over the machines and they run in containers, which is usually the case. This is not a possibility.
  
The contributing guide clearly states that any change touching the core logic of the incremental compiler has to go through benchmarks. For you, this may be uninteresting or unnecessary. For me, as now maintainer of Zinc, it's pretty serious.
  
The fact that I want to get predictable benchmarking results and get things right should not hinder a more in-depth analysis on the consequences of this PR.
  
It's true that the author of this PR should perhaps not be involved in it. So I'll take care of this from now on. But I'm swamped in work, so this may delay some days more.

smarter commented:
Has anyone done any profiling of zinc at all, using VisualVM or YourKit Profiler? I suspect you'll find much more improvements to be made from figuring out what to optimize based on profiling rather than relying on macro benchmarks results.

jvican commented:
> Has anyone done any profiling of zinc at all, using VisualVM or YourKit Profiler? I suspect the time you'll find much more improvement to be made from figuring out what to optimize based on profiling rather than relying on macro benchmarks results.
  
Yes, I have, and it's not easy, particularly because scalac results intermingle with Zinc's. I think we need to improve our benchmarking suite and find out a way to attach Yourkit hooks in the beginning of the phases. But this requires time and it's not a priority for me now.

romanowski commented:
@smarter my recent speedups was all based on Yourkit output. Baenchmarks are good to find any big breaks (but situation can be improved if our benchmarks produce times for zinc related phases alongside with full time).

@jvican I usually profile whole build and later on narrow down (in YourKit 'focus on method') scope to given phase.

ttreyer commented:
I ran the benchmarks again following your indications, here are the results: 
[benchmarks-before.txt](https://github.com/sbt/zinc/files/850590/benchmarks-before.txt)
[benchmarks-after.txt](https://github.com/sbt/zinc/files/850592/benchmarks-after.txt)

Sorry for not responding earlier, but I can only run the benchmarks overnight when I don't need my laptop.

dwijnand reviewed (dismissed):
I'm "requesting changes" just to indicate in the GitHub UI that still needs attention WRT the performance (see latest comments).

jvican commented:
I'll get back to this one soon.

jvican commented:
I've run benchmarks on my machine because @zehkae's results were suspiciously unreasonable. I don't know how the results were showing such a negative performance impact of this change.
  
BEFORE:
```
[info] # Run complete. Total time: 00:25:57
[info] 
[info] Benchmark                                                           (_tempDir)    Mode  Cnt           Score            Error   Units
[info] HotScalacBenchmark.compile                                    /tmp/sbt_919ba19  sample   18       20721.726 ±        665.170   ms/op
[info] HotScalacBenchmark.compile:compile·p0.00                      /tmp/sbt_919ba19  sample            19528.679                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.50                      /tmp/sbt_919ba19  sample            20854.079                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.90                      /tmp/sbt_919ba19  sample            21635.898                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.95                      /tmp/sbt_919ba19  sample            21877.490                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.99                      /tmp/sbt_919ba19  sample            21877.490                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.999                     /tmp/sbt_919ba19  sample            21877.490                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999                    /tmp/sbt_919ba19  sample            21877.490                    ms/op
[info] HotScalacBenchmark.compile:compile·p1.00                      /tmp/sbt_919ba19  sample            21877.490                    ms/op
[info] HotScalacBenchmark.compile:·gc.alloc.rate                     /tmp/sbt_919ba19  sample   18         291.385 ±         10.283  MB/sec
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                /tmp/sbt_919ba19  sample   18  6483756999.556 ±   38074179.544    B/op
[info] HotScalacBenchmark.compile:·gc.churn.PS_Eden_Space            /tmp/sbt_919ba19  sample   18         288.506 ±         23.916  MB/sec
[info] HotScalacBenchmark.compile:·gc.churn.PS_Eden_Space.norm       /tmp/sbt_919ba19  sample   18  6423663957.333 ±  537833505.454    B/op
[info] HotScalacBenchmark.compile:·gc.churn.PS_Old_Gen               /tmp/sbt_919ba19  sample   18          13.500 ±         13.043  MB/sec
[info] HotScalacBenchmark.compile:·gc.churn.PS_Old_Gen.norm          /tmp/sbt_919ba19  sample   18   307806890.667 ±  297145775.885    B/op
[info] HotScalacBenchmark.compile:·gc.churn.PS_Survivor_Space        /tmp/sbt_919ba19  sample   18           6.972 ±          2.564  MB/sec
[info] HotScalacBenchmark.compile:·gc.churn.PS_Survivor_Space.norm   /tmp/sbt_919ba19  sample   18   156339664.444 ±   59747847.126    B/op
[info] HotScalacBenchmark.compile:·gc.count                          /tmp/sbt_919ba19  sample   18         102.000                   counts
[info] HotScalacBenchmark.compile:·gc.time                           /tmp/sbt_919ba19  sample   18       22066.000                       ms
[info] WarmScalacBenchmark.compile                                   /tmp/sbt_919ba19  sample    3       53284.438 ±      39804.378   ms/op
[info] WarmScalacBenchmark.compile:compile·p0.00                     /tmp/sbt_919ba19  sample            51069.846                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.50                     /tmp/sbt_919ba19  sample            53351.547                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.90                     /tmp/sbt_919ba19  sample            55431.922                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.95                     /tmp/sbt_919ba19  sample            55431.922                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.99                     /tmp/sbt_919ba19  sample            55431.922                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.999                    /tmp/sbt_919ba19  sample            55431.922                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.9999                   /tmp/sbt_919ba19  sample            55431.922                    ms/op
[info] WarmScalacBenchmark.compile:compile·p1.00                     /tmp/sbt_919ba19  sample            55431.922                    ms/op
[info] WarmScalacBenchmark.compile:·gc.alloc.rate                    /tmp/sbt_919ba19  sample    3         123.283 ±         99.933  MB/sec
[info] WarmScalacBenchmark.compile:·gc.alloc.rate.norm               /tmp/sbt_919ba19  sample    3  7006538144.000 ±  565872164.065    B/op
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Eden_Space           /tmp/sbt_919ba19  sample    3         116.214 ±         71.571  MB/sec
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Eden_Space.norm      /tmp/sbt_919ba19  sample    3  6608815581.333 ± 3863172275.767    B/op
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Survivor_Space       /tmp/sbt_919ba19  sample    3           2.445 ±          4.584  MB/sec
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Survivor_Space.norm  /tmp/sbt_919ba19  sample    3   138730165.333 ±  166744646.238    B/op
[info] WarmScalacBenchmark.compile:·gc.count                         /tmp/sbt_919ba19  sample    3          75.000                   counts
[info] WarmScalacBenchmark.compile:·gc.time                          /tmp/sbt_919ba19  sample    3        8668.000                       ms
[info] ColdScalacBenchmark.compile                                   /tmp/sbt_919ba19      ss   10       44980.828 ±       1047.047   ms/op
[info] ColdScalacBenchmark.compile:·gc.alloc.rate                    /tmp/sbt_919ba19      ss   10         149.075 ±          3.742  MB/sec
[info] ColdScalacBenchmark.compile:·gc.alloc.rate.norm               /tmp/sbt_919ba19      ss   10  7174553936.000 ±   94085063.199    B/op
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Eden_Space           /tmp/sbt_919ba19      ss   10         140.662 ±          6.393  MB/sec
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Eden_Space.norm      /tmp/sbt_919ba19      ss   10  6770402511.200 ±  314960053.291    B/op
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Old_Gen              /tmp/sbt_919ba19      ss   10           0.002 ±          0.008  MB/sec
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Old_Gen.norm         /tmp/sbt_919ba19      ss   10      109006.400 ±     368292.066    B/op
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Survivor_Space       /tmp/sbt_919ba19      ss   10           2.996 ±          0.345  MB/sec
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Survivor_Space.norm  /tmp/sbt_919ba19      ss   10   144192171.200 ±   16592896.254    B/op
[info] ColdScalacBenchmark.compile:·gc.count                         /tmp/sbt_919ba19      ss   10         252.000                   counts
[info] ColdScalacBenchmark.compile:·gc.time                          /tmp/sbt_919ba19      ss   10       29289.000                       ms
[success] Total time: 1559 s, completed Apr 4, 2017 1:41:00 PM
[success] Total time: 0 s, completed Apr 4, 2017 1:41:00 PM
```
  
AFTER:
```
[info] # Run complete. Total time: 00:25:59
[info] 
[info] Benchmark                                                            (_tempDir)    Mode  Cnt           Score            Error   Units
[info] HotScalacBenchmark.compile                                    /tmp/sbt_1eb81994  sample   18       21021.852 ±        584.833   ms/op
[info] HotScalacBenchmark.compile:compile·p0.00                      /tmp/sbt_1eb81994  sample            19964.887                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.50                      /tmp/sbt_1eb81994  sample            21189.624                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.90                      /tmp/sbt_1eb81994  sample            21917.755                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.95                      /tmp/sbt_1eb81994  sample            21978.153                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.99                      /tmp/sbt_1eb81994  sample            21978.153                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.999                     /tmp/sbt_1eb81994  sample            21978.153                    ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999                    /tmp/sbt_1eb81994  sample            21978.153                    ms/op
[info] HotScalacBenchmark.compile:compile·p1.00                      /tmp/sbt_1eb81994  sample            21978.153                    ms/op
[info] HotScalacBenchmark.compile:·gc.alloc.rate                     /tmp/sbt_1eb81994  sample   18         287.482 ±          8.089  MB/sec
[info] HotScalacBenchmark.compile:·gc.alloc.rate.norm                /tmp/sbt_1eb81994  sample   18  6488495421.333 ±   37767637.669    B/op
[info] HotScalacBenchmark.compile:·gc.churn.PS_Eden_Space            /tmp/sbt_1eb81994  sample   18         291.150 ±         22.934  MB/sec
[info] HotScalacBenchmark.compile:·gc.churn.PS_Eden_Space.norm       /tmp/sbt_1eb81994  sample   18  6570901504.000 ±  474416737.126    B/op
[info] HotScalacBenchmark.compile:·gc.churn.PS_Old_Gen               /tmp/sbt_1eb81994  sample   18          15.226 ±         12.359  MB/sec
[info] HotScalacBenchmark.compile:·gc.churn.PS_Old_Gen.norm          /tmp/sbt_1eb81994  sample   18   349794773.778 ±  282991988.622    B/op
[info] HotScalacBenchmark.compile:·gc.churn.PS_Survivor_Space        /tmp/sbt_1eb81994  sample   18           7.021 ±          2.916  MB/sec
[info] HotScalacBenchmark.compile:·gc.churn.PS_Survivor_Space.norm   /tmp/sbt_1eb81994  sample   18   159634081.778 ±   68370763.968    B/op
[info] HotScalacBenchmark.compile:·gc.count                          /tmp/sbt_1eb81994  sample   18         106.000                   counts
[info] HotScalacBenchmark.compile:·gc.time                           /tmp/sbt_1eb81994  sample   18       23959.000                       ms
[info] WarmScalacBenchmark.compile                                   /tmp/sbt_1eb81994  sample    3       53105.481 ±       8333.743   ms/op
[info] WarmScalacBenchmark.compile:compile·p0.00                     /tmp/sbt_1eb81994  sample            52747.567                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.50                     /tmp/sbt_1eb81994  sample            52948.894                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.90                     /tmp/sbt_1eb81994  sample            53619.982                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.95                     /tmp/sbt_1eb81994  sample            53619.982                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.99                     /tmp/sbt_1eb81994  sample            53619.982                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.999                    /tmp/sbt_1eb81994  sample            53619.982                    ms/op
[info] WarmScalacBenchmark.compile:compile·p0.9999                   /tmp/sbt_1eb81994  sample            53619.982                    ms/op
[info] WarmScalacBenchmark.compile:compile·p1.00                     /tmp/sbt_1eb81994  sample            53619.982                    ms/op
[info] WarmScalacBenchmark.compile:·gc.alloc.rate                    /tmp/sbt_1eb81994  sample    3         123.700 ±         21.462  MB/sec
[info] WarmScalacBenchmark.compile:·gc.alloc.rate.norm               /tmp/sbt_1eb81994  sample    3  7017565357.333 ±  296877411.797    B/op
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Eden_Space           /tmp/sbt_1eb81994  sample    3         113.270 ±         69.536  MB/sec
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Eden_Space.norm      /tmp/sbt_1eb81994  sample    3  6425366538.667 ± 3358459955.066    B/op
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Survivor_Space       /tmp/sbt_1eb81994  sample    3           2.648 ±          2.103  MB/sec
[info] WarmScalacBenchmark.compile:·gc.churn.PS_Survivor_Space.norm  /tmp/sbt_1eb81994  sample    3   150189434.667 ±  104657211.705    B/op
[info] WarmScalacBenchmark.compile:·gc.count                         /tmp/sbt_1eb81994  sample    3          76.000                   counts
[info] WarmScalacBenchmark.compile:·gc.time                          /tmp/sbt_1eb81994  sample    3        9262.000                       ms
[info] ColdScalacBenchmark.compile                                   /tmp/sbt_1eb81994      ss   10       44784.657 ±        967.608   ms/op
[info] ColdScalacBenchmark.compile:·gc.alloc.rate                    /tmp/sbt_1eb81994      ss   10         149.636 ±          3.113  MB/sec
[info] ColdScalacBenchmark.compile:·gc.alloc.rate.norm               /tmp/sbt_1eb81994      ss   10  7173049382.400 ±   71999309.520    B/op
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Eden_Space           /tmp/sbt_1eb81994      ss   10         139.034 ±          8.819  MB/sec
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Eden_Space.norm      /tmp/sbt_1eb81994      ss   10  6664568044.800 ±  393275917.747    B/op
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Survivor_Space       /tmp/sbt_1eb81994      ss   10           3.095 ±          0.893  MB/sec
[info] ColdScalacBenchmark.compile:·gc.churn.PS_Survivor_Space.norm  /tmp/sbt_1eb81994      ss   10   148218612.000 ±   41195067.353    B/op
[info] ColdScalacBenchmark.compile:·gc.count                         /tmp/sbt_1eb81994      ss   10         250.000                   counts
[info] ColdScalacBenchmark.compile:·gc.time                          /tmp/sbt_1eb81994      ss   10       28821.000                       ms
```
  
The results look much better! :smile:.
  
In short, this PR introduces a slowdown of 100ms while hot in p95, while decreasing the margin error for 80ms for the average case. While warm, this PR is faster by almost 2 seconds (and margin error is dramatically decreased for the average case). Lastly, while cold this PR brings 200ms improvement. There are almost no difference in memory consumption.
  
After all, this PR does not seem to affect negatively the performance of Zinc. Given my results, I feel pretty confident merging this.
  
Thank you for all your work @zekhae. @smarter Hope this also helps you on the Dotty side.
